pkg/controller/custom_resource_replicas.go

```diff
@@ -0,0 +1,150 @@
+		}
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func getCustomScaleResourceForStatefulset(ctx context.Context, sts *appsv1.StatefulSet, restMapper meta.RESTMapper, scalesGetter scaleclient.ScalesGetter) (*autoscalingv1.Scale, schema.GroupVersionResource, string, error) {
+	annotations := sts.GetAnnotations()
+	name := annotations[config.RolloutMirrorReplicasFromResourceNameAnnotationKey]
+	kind := annotations[config.RolloutMirrorReplicasFromResourceKindAnnotationKey]
+	if name == "" || kind == "" {
+		return nil, schema.GroupVersionResource{}, "", nil
+	}
+
+	apiVersion := annotations[config.RolloutMirrorReplicasFromResourceAPIVersionAnnotationKey]
+
+	targetGV, err := schema.ParseGroupVersion(apiVersion)
+	if err != nil {
+		return nil, schema.GroupVersionResource{}, "", fmt.Errorf("invalid API version in %s annotation: %v", config.RolloutMirrorReplicasFromResourceAPIVersionAnnotationKey, err)
+	}
+
+	targetGK := schema.GroupKind{
+		Group: targetGV.Group,
+		Kind:  kind,
+	}
+
+	reference := fmt.Sprintf("%s/%s", kind, name)
+
+	mappings, err := restMapper.RESTMappings(targetGK)
+	if err != nil {
+		return nil, schema.GroupVersionResource{}, "", fmt.Errorf("unable to find custom resource mapping for reference resource %s: %v", reference, err)
+	}
+
+	scale, gvr, err := scaleForResourceMappings(ctx, sts.Namespace, name, mappings, scalesGetter)
+	if err != nil {
+		return nil, schema.GroupVersionResource{}, "", fmt.Errorf("failed to query scale subresource for %s: %v", reference, err)
+	}
+
+	return scale, gvr, name, nil
+}
+
+// copied from https://github.com/kubernetes/kubernetes/blob/3c4512c6ccca066d590a33b6333198b5ed813da2/pkg/controller/podautoscaler/horizontal.go#L1336-L1358
+func scaleForResourceMappings(ctx context.Context, namespace, name string, mappings []*meta.RESTMapping, scalesGetter scaleclient.ScalesGetter) (*autoscalingv1.Scale, schema.GroupVersionResource, error) {
+	var firstErr error
+	for i, mapping := range mappings {
+		scale, err := scalesGetter.Scales(namespace).Get(ctx, mapping.Resource.GroupResource(), name, metav1.GetOptions{})
+		if err == nil {
+			return scale, mapping.Resource, nil
+		}
+
+		// if this is the first error, remember it,
+		// then go on and try other mappings until we find a good one
+		if i == 0 {
+			firstErr = err
+		}
+	}
+
+	// make sure we handle an empty set of mappings
+	if firstErr == nil {
+		firstErr = fmt.Errorf("unrecognized resource")
+	}
+
+	return nil, schema.GroupVersionResource{}, firstErr
+}
+
+// updateStatusReplicasOnMirroredResourceIfNeeded makes sure that scaleObject's status.replicas field is up-to-date.
+// if update fails, error is logged, but not returned to caller.
+func updateStatusReplicasOnMirroredResourceIfNeeded(ctx context.Context, log log.Logger, dynamicClient dynamic.Interface, sts *appsv1.StatefulSet, scaleObj *autoscalingv1.Scale, gvr schema.GroupVersionResource, resName string, replicas int32) {
```

Comment:
[nit] Should we rename this too?

```suggestion
func updateStatusReplicasOnReferenceResourceIfNeeded(ctx context.Context, log log.Logger, dynamicClient dynamic.Interface, sts *appsv1.StatefulSet, scaleObj *autoscalingv1.Scale, gvr schema.GroupVersionResource, resName string, replicas int32) {
```